proxy/Dockerfile

```diff
@@ -1,24 +1,106 @@
 # Proxy build and runtime
 #
-# Builds a slim runtime image with the conduit-proxy binary.
+# When BUILD_UNOPTIMIZED is set and not empty, unoptimized rust artifacts are produced.
+# This reduces build time and produces binaries with debug symbols, at the expense of
+# runtime perforamnce.
+#
+# When SKIP_TESTS is set and not empty, tests are not run. Otherwise, tests are run
+# against either unoptimized or optimized proxy code, according to BUILD_UNOPTIMIZED.
+
+ARG RUST_IMAGE=rust:1.23.0
+
+ARG RUNTIME_IMAGE=gcr.io/runconduit/base:2017-10-30.01
+
+## Fetch external dependencies.
+#
+# Mocks out all local code and fetch external dependencies to ensure that external sources
+# are cached.
+FROM $RUST_IMAGE as fetch
+WORKDIR /usr/src/conduit
+RUN for d in proxy proxy/controller-grpc proxy/convert proxy/futures-mpsc-lossy proxy/router ; \
+    do mkdir -p "${d}/src" && touch "${d}/src/lib.rs" ; \
+    done
+COPY Cargo.toml Cargo.lock ./
+COPY proxy/Cargo.toml                       proxy/Cargo.toml
+COPY proxy/controller-grpc/Cargo.toml       proxy/controller-grpc/Cargo.toml
+COPY proxy/convert/Cargo.toml               proxy/convert/Cargo.toml
+COPY proxy/futures-mpsc-lossy/Cargo.toml    proxy/futures-mpsc-lossy/Cargo.toml
+COPY proxy/router/Cargo.toml                proxy/router/Cargo.toml
+RUN cargo fetch --locked
+
+## Build libraries.
+#
+# Mocks proxy sources and gRPC bindings to ensure that dependencies are cached.
+FROM $RUST_IMAGE as libs
+WORKDIR /usr/src/conduit
+RUN for d in proxy proxy/controller-grpc ; \
+    do mkdir -p "${d}/src" && touch "${d}/src/lib.rs" ; \
+    done
+COPY Cargo.toml Cargo.lock              ./
+COPY proxy/Cargo.toml                   ./proxy/Cargo.toml
+COPY proxy/controller-grpc/Cargo.toml   ./proxy/controller-grpc/Cargo.toml
+COPY proxy/convert                      ./proxy/convert
+COPY proxy/futures-mpsc-lossy           ./proxy/futures-mpsc-lossy
+COPY proxy/router                       ./proxy/router
+COPY --from=fetch $CARGO_HOME $CARGO_HOME
+ARG BUILD_OPTIMIZED
+RUN if [ -z "$BUILD_OPTIMIZED" ]; \
+    then cargo build --frozen ; \
+    else cargo build --frozen --release ; \
+    fi
 
-## Build the rust proxy into a binary.
+## Build gRPC bindings.
 #
-# If the RELEASE arg is set and non-empty, a release artifact is built.
-FROM gcr.io/runconduit/proxy-deps:673c53de as build
+# Mocks proxy sources to ensure that dependencies, including build-time dependencies
+# needed to generate the bindings, are cached.
+FROM $RUST_IMAGE as grpc
```

Comment:
Please add a comment about why this is a separate named stage, instead of just being the end of the previous stage or the beginning of the next stage. (AFAICT Docker will cache things about the same whichever way is chosen.)